OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -165,11 +174,64 @@ static bool IsMovingInMyDirection(Actor self, Actor other)
 			return true;
 		}
 
+		public int TileSetMovementHash(TileSet tileSet)
+		{
+			var a = TilesetTerrainInfo[tileSet];
+			int hc = a.Length;
+			for (int i = 0; i < a.Length; i++)
+			{
+				hc = unchecked(hc * 31 + a[i].Cost);
+			}
+
+			return hc;
+		}
+
 		public bool CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor = null, CellConditions check = CellConditions.All)
 		{
 			if (MovementCostForCell(world, cell) == int.MaxValue)
 				return false;
 
+			return CanMoveFreely(world, self, cell, ignoreActor, check);
+		}
+
+		// Determines whether the actor is blocked by other Actors
+		public bool CanMoveFreely(World world, Actor self, CPos cell, Actor ignoreActor, CellConditions check)
```

Comment:
This new name doesn't sit well with me. `CanMoveFreelyInto` or similar (the `CanEnterCell` was fine?) illustrates more clearly what this test, IMO.